doc/control_mechanisms/petri_nets.md

```diff
@@ -0,0 +1,192 @@
+- Für jede Aufgabe existiert eine Stelle, die (falls markiert) angibt, ob die Aufgabe aktuell zur Bearbeitung
+  freigeschaltet ist
+- Für jede Aufgabe existiert eine Stelle, die (falls markiert) angibt, ob die Aufgabe abgeschlossen ist
+- Es muss eine Schnittstelle zum Dungeon / zur DSL vorgesehen werden, über die dem Petri-Netz mitgeteilt
+  werden kann, dass eine Aufgabe bearbeitet wurde. Diese Information wird nicht vom Petri-Netz selbst abgefragt,
+  sondern vom Dungeon / der DSL an das Petri-Netz übermittelt, da dieser Prozess eng mit dem konkreten Szenario
+  verknüpft ist, welches in der DSL definiert ist.
+
+
+### Erforderliche Teilaufgabe
+
+![Erforderliche Teilaufgabe: Graph](img/mandatory_subtask.png)
+
+Für $t_2$ muss eine Antwort abgegeben werden, bevor $t_1$ abgeschlossen werden kann;
+$t_1$ wird zuerst aktiviert und bleibt aktiv, während $t_2$ bearbeitet wird.
+
+Kürzel: "subtask_mandatory"/"st_m"
+
+Eingabe:
+```
+task_dependency t {
+  t1 -> t2 [type="st_m"]
+}
+```
+
+Das erzeugte Petri-Netz für diese Beziehung sieht wie folgt aus:
+
+![Erforderliche Teilaufgabe: Petri-Netz](img/mandatory_subtask_petri.png)
+
+Die Transitionen, die von links und rechts in das Petri-Netz übergehen (z.B. "t1 aktivieren", "t2 abschließen"),
+sind die Schnittstellen zum Dungeon / zur DSL.
+
+Der Zustand einer Aufgabe (z.B. $t_1$) kann aus der Kombination der Stellen "t1 Bearb. aktiviert" und "t1 abgeschlossen"
+hergeleitet werden. In den Transitionen können Aktionen ausgeführt werden, wie z.B. eine Aufgabe im Questlog anzuzeigen,
+oder sie aus dem Questlog zu löschen. Der Zustand **aktiv ohne Bearbeitung** wird daher nicht explizit im Petri-Netz
+modelliert.
+
+Es folgt:
+
+- Stelle "t1 Bearb. aktiviert" und "t1 abgeschlossen" sind beide **nicht markiert** = Zustand **inaktiv**
+- Stelle "t1 Bearb. aktiviert" ist **markiert** und "t1 abgeschlossen" ist **nicht markiert** = Zustand **aktiv mit Bearbeitung**
+- Stelle "t1 Bearb. aktiviert" ist **nicht markiert** und "t1 abgeschlossen" ist **markiert** = Zustand **fertig bearbeitet**
+
+Eine Aufgabe kann mehrere erforderliche Teilaufgaben haben, wie im folgenden Graph dargestellt:
+
+![Mehrere erforderliche Teilaufgaben: Graph](img/multi_mandatory_subtask.png)
+
+Das entsprechende Petri-Netz sieht so aus:
+
+![Mehrere erforderliche Teilaufgaben: Petri-Netz](img/multi_mandatory_subtask_petri.png)
+
+Der grün markierte Bereich dient zur Behandlung von $t_2$, der blau markierte Bereich dient zur Behandlung von $t_3$.
+Weitere Teilaufgaben würden auf der gleichen Ebene mit der gleichen Petri-Netz-Struktur hinzugefügt werden.
+Der rot markierte Bereich dient zur Überprüfung, ob beide Teilaufgaben abgeschlossen sind, bevor die Bearbeitung von $t_1$
+freigeschaltet wird. Die Transition in diesem Bereich konsumiert jeweils ein Token von den Stellen "t2 abgeschlossen" und
+"t3 abgeschlossen", daher werden diese Stellen jeweils mit 2 Tokens markiert (wie an deren eingehenden Kanten durch "+2"
```

Comment:
Wieso konsumiert der Übergang nach "t2 abgeschlossen" bzw. "t3 abgeschlossen" jeweils 2 Token? Müsste das dann nicht oben im ersten Petrinetz auch so sein?

Reply:
Das hängt damit zusammen, dass die "Bearbeitung aktiviert"- und "tx abgeschlossen"-Stellen vom Dungeon ausgelesen werden sollen, um den Zustand der Aufgaben zu ermitteln.
Im oberen Petri-Netz ist keine Transition von der "t2 abgeschlossen"-Stelle abhängig, also reicht ein Token, um die Stelle zu markieren. In dem Petri-Netz, was hier beschrieben ist, ist eine Transition von "t2 abgeschlossen" und "t3 abgeschlossen" abhängig, und wird ein Token konsumieren, sobald sie schaltet. Um auch nach dem Schalten noch zuverlässig den Zustand von t2 und t3 aus dem Petri-Netz auslesen zu können, werden da initial jeweils 2 Tokens in die Stellen geschoben.
Hab ich auch so im Text ergänzt.

Reply:
@malt-r Das verstehe ich noch nicht. 

Normalerweise kann eine Transition schalten, wenn in allen Eingangsstellen einer Transition mind. je ein Token enthalten ist (wenn man nix anderes dran schreibt). Dann nimmt sie je ein Token dort raus, und in die Ausgangsstellen wird je ein Token reingelegt. Das ist in allen Bildern bei Dir so.

In dieser Abbildung mit mehreren mandatory subtasks habe ich im roten Block zwei Stellen, die jeweils aus dem grünen bzw. blauen Block davor "gefüttert" werden. Wenn einer von den beiden schaltet, würden jetzt zwei Token nach "t2 abgeschlossen" oder "t3 abgeschlossen" gehen. Das nutzt mir aber nichts, weil ich in der jeweils anderen Stelle auch eine Marke brauche, um nach "t1 Bearb. aktiviert" weiter gehen zu können. 

Irgendwas blicke ich nicht.

Reply:
Alter. Jetzt. Hmmm.

Du hast ein Petri-Netz - wenn Du mit je 2 Markierungen nach "t2 abgeschlossen" und "t3 abgeschlossen" gehst, dann kann die folgende Transition 2x schalten - und Du kannst im Dungeon nix auslesen.

Mein Vorschlag wäre, analog zu den Eingängen aus dem Dungeon explizite Ausgänge (Pfeil von der Stelle zu einem kleinen schwarzen Rechteck) in den Dungeon zu modellieren und dann auch eine extra Stelle zu spendieren.

Reply:
Alter. Jetzt. Hmmm.

Du hast ein Petri-Netz - wenn Du mit je 2 Markierungen nach "t2 abgeschlossen" und "t3 abgeschlossen" gehst, dann kann die folgende Transition 2x schalten - und Du kannst im Dungeon nix auslesen. Ist irgendwie beides nicht das, was Du hier willst.

Mein Vorschlag wäre, analog zu den Eingängen aus dem Dungeon explizite Ausgänge (Pfeil von der Stelle zu einem kleinen schwarzen Rechteck) in den Dungeon zu modellieren und dann auch eine extra Stelle zu spendieren.

@bcg7 Was hältst Du davon?

Reply:
Ups, stimmt! Dass die Transition zweimal schalten kann, habe ich nicht bedacht. Die Ausgangs-Stellen/Transitionen explizit zu modellieren ist wahrscheinlich zielführender.

Reply:
Ich habe die Ausgangstrasitionen in den Petri-Netzen hinzugefügt und die Texte angepasst.